vnext/Shared/MemoryMappedBuffer.cpp

```diff
@@ -0,0 +1,154 @@
+    : m_fileMapping{nullptr, &CloseHandle},
+      m_fileData{nullptr, &fileDataDeleter},
+      m_size{0},
+      m_offset{0} {
+#if (defined(WINRT))
+  std::unique_ptr<void, decltype(&CloseHandle)> fileHandle{
+      CreateFile2(
+          filename,
+          GENERIC_READ,
+          FILE_SHARE_READ,
+          OPEN_EXISTING,
+          nullptr /* pCreateExParams */),
+      &CloseHandle};
+#else
+  std::unique_ptr<void, decltype(&CloseHandle)> fileHandle{
+      CreateFileW(
+          filename,
+          GENERIC_READ,
+          FILE_SHARE_READ,
+          nullptr /* lpSecurityAttributes */,
+          OPEN_EXISTING,
+          FILE_ATTRIBUTE_NORMAL,
+          nullptr /* hTemplateFile */),
+      &CloseHandle};
+#endif
+
+  if (fileHandle.get() == INVALID_HANDLE_VALUE) {
+    throw facebook::jsi::JSINativeException(
+        "CreateFileW/CreateFile2 failed with last error " +
+        std::to_string(GetLastError()));
+  }
+
+  LARGE_INTEGER size;
+  if (!GetFileSizeEx(fileHandle.get(), &size)) {
+    throw facebook::jsi::JSINativeException(
+        "GetFileSizeEx failed with last error " +
+        std::to_string(GetLastError()));
+  }
+
+  if (size.LowPart == 0) {
+    throw facebook::jsi::JSINativeException("Cannot memory map an empty file.");
+  }
+
+  if (size.HighPart != 0) {
+    throw facebook::jsi::JSINativeException(
+        "MemoryMappedBuffer only support files whose size can fit within an "
+        "uint32_t.");
+  }
+
+  m_size = size.LowPart;
+  if (offset > m_size) {
+    m_size = 0;
+    throw facebook::jsi::JSINativeException("Invalid offset.");
+  }
+  m_offset = offset;
+
+#if (defined(WINRT))
+  m_fileMapping.reset(CreateFileMappingFromApp(
+      fileHandle.get(),
+      nullptr /* SecurityAttributes */,
+      PAGE_READONLY,
+      m_size,
+      nullptr /* Name */));
+#else
+  m_fileMapping.reset(CreateFileMapping(
+      fileHandle.get(),
+      nullptr /* lpAttributes */,
+      PAGE_READONLY,
+      0 /* dwMaximumSizeHigh */,
+      m_size,
+      nullptr /* lpName */));
+#endif
+
+  if (!m_fileMapping) {
+    m_size = m_offset = 0;
+    throw facebook::jsi::JSINativeException(
+        "CreateFileMapping/CreateFileMappingFromApp failed with last error " +
+        std::to_string(GetLastError()));
+  }
+
+#if (defined(WINRT))
+  m_fileData.reset(MapViewOfFileFromApp(
+      m_fileMapping.get(),
+      FILE_MAP_READ,
+      0 /* FileOffset */,
+      0 /* NumberOfBytesToMap */));
+#else
+  m_fileData.reset(MapViewOfFile(
+      m_fileMapping.get(),
+      FILE_MAP_READ,
+      0 /* dwFileOffsetHigh */,
+      0 /* dwFileOffsetLow */,
+      0 /* dwNumberOfBytesToMap */));
+#endif
+
+  if (!m_fileData) {
+    m_size = m_offset = 0;
+    throw facebook::jsi::JSINativeException(
+        "MapViewOfFile/MapViewOfFileFromApp failed with last error " +
+        std::to_string(GetLastError()));
+  }
+
+  WerRegisterMemoryBlock(m_fileData.get(), m_size);
+}
+
+size_t MemoryMappedBuffer::size() const {
```

Comment:
>size_t [](start = 0, length = 6)

it would be nicer to keep consistent types, either size_t or uint32_t

Reply:
This method is inherited from facebook::jsi::Buffer, so its return type has to be size_t. I tried to use uint32_t everywhere else because uint32_t's size is guaranteed.

---
In reply to: [320912391](https://github.com/microsoft/react-native-windows/pull/3059#discussion_r320912391) [](ancestors = 320912391)